controllers/topology/internal/extensions/patches/engine.go

```diff
@@ -0,0 +1,495 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package patches implement the patch engine.
+package patches
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/pkg/errors"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers/topology/internal/contract"
+	"sigs.k8s.io/cluster-api/controllers/topology/internal/extensions/patches/api"
+	"sigs.k8s.io/cluster-api/controllers/topology/internal/extensions/patches/variables"
+	tlog "sigs.k8s.io/cluster-api/controllers/topology/internal/log"
+	"sigs.k8s.io/cluster-api/controllers/topology/internal/scope"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Apply applies patches to the desired state according to the patches from the ClusterClass, variables from the Cluster
+// and builtin variables.
+// * A GenerateRequest with all templates and global and template-specific variables is created.
+// * Then for all patches of a ClusterClass, patches are generated and successively applied to the templates
+//   in the GenerateRequest.
+// * Eventually the patched templates are used to update the specs of the desired objects.
+func Apply(ctx context.Context, blueprint *scope.ClusterBlueprint, desired *scope.ClusterState) error {
+	log := tlog.LoggerFrom(ctx)
+
+	// Create a patch generation request.
+	req, err := request(blueprint, desired)
+	if err != nil {
+		return errors.Wrapf(err, "failed to generate patch request")
+	}
+
+	// Loop over patches in ClusterClass, generate patches and apply them to the request.
+	for i := range blueprint.ClusterClass.Spec.Patches {
+		patch := blueprint.ClusterClass.Spec.Patches[i]
```

Comment:
```suggestion
		patchDefinition := blueprint.ClusterClass.Spec.Patches[i]
```
To disambiguate from the generated patch

Reply:
Not sure if we have to disambiguate as the generated patch is only used inside of `applyPatchesToRequest` and I think it's better to keep the var names short versus trying to keep the var names unique per struct.

(I would have to name the var `clusterClassPatch` as that's the corresponding struct)

Reply:
Not blocking, but as it is not it read that we have a patch and then we use it to generated other patches, and this could be confusing at a first sight. this is why I suggested a more verbose name

Reply:
Okay that makes sense. I thought it's about using different variable names for different structs in general as brought up here https://github.com/kubernetes-sigs/cluster-api/pull/5534#issuecomment-957030201

Reply:
I'll call it `clusterClassPatch` though to avoid the impression that it's a `PatchDefinition`.